grudge/dt_utils.py

```diff
@@ -0,0 +1,206 @@
+"""Helper functions for estimating stable time steps for RKDG methods.
+
+.. autofunction:: dt_non_geometric_factor
+.. autofunction:: dt_geometric_factor
+"""
+
+__copyright__ = """
+Copyright (C) 2021 University of Illinois Board of Trustees
+"""
+
+__license__ = """
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in
+all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+"""
+
+
+import numpy as np
+
+from arraycontext import make_loopy_program
+
+from grudge.dof_desc import DD_VOLUME, DOFDesc
+from grudge.discretization import DiscretizationCollection
+
+import grudge.op as op
+
+from meshmode.dof_array import DOFArray
+
+from pytools import memoize_on_first_arg, memoize_in
+
+
+@memoize_on_first_arg
+def dt_non_geometric_factor(
+        dcoll: DiscretizationCollection, scaling=None, dd=None) -> float:
+    r"""Computes the non-geometric scale factor:
+
+    .. math::
+
+        C\operatorname{min}_i\left( \Delta r_i \right),
+
+    where :math:`\Delta r_i` denotes the distance between two distinct
+    nodes on the reference element, and :math:`C > 0` is a scaling
+    constant.
+
+    :arg scaling: a :class:`float` denoting the scaling factor. By default,
+        the constant is set to 2/3.
```

Comment:
I'd kind of feel better if we removed this argument and let users scale the returned time step. Otherwise we end up documenting the 2/3, which feels kinda  weird. (But I'm open to discussion here.)

Reply:
This is based on Hesthaven and Warburton, which is heuristically determined.

@MTCam this means in your code, you'll need to add a factor of `2/3` in front of `non_geometric_factor * geometric_factor` in your timestep calculation. Does that feel okay with you?

Reply:
Marked to address this in a separate PR: https://github.com/inducer/grudge/pull/108/commits/f18a36970bd18dfe7e890d19813d97b9f0aed5ab